packages/pam/handlers/mysql/relay_handler.go

```diff
@@ -114,6 +143,43 @@ func (r *RelayHandler) writeLogEntry(entry session.SessionLogEntry) (*mysql.Resu
 	return nil, nil
 }
 
+func (r *RelayHandler) handleReadOnlyCheck(query string) (bool, error) {
+	if query == "" {
+		return true, nil
+	}
+
+	dialect := sqi.DialectMySQL
+	strict := false
+	options := sqi.IdentifyOptions{
+		Dialect: &dialect,
+		Strict:  &strict,
+	}
+
+	identifiedQueries, err := sqi.Identify(query, options)
+	if err != nil {
+		log.Error().
+			Str("sessionID", r.config.SessionID).
+			Str("query", query).
+			Err(err).
+			Msg("Failed to identify query; blocking in read-only mode.")
+		return false, err
+	}
+
+	// verify that every statement in the query is read-only
+	for _, identifiedQuery := range identifiedQueries {
+		if identifiedQuery.ExecutionType != sqi.ExecutionListing && identifiedQuery.ExecutionType != sqi.ExecutionInformation {
+			log.Warn().
+				Str("sessionID", r.config.SessionID).
+				Str("query", query).
+				Str("executionType", string(identifiedQuery.ExecutionType)).
```

Comment:
There should be automatic unit test tests covering different SQL statements trying to see if it can parse them correctly and detect write syntax. One test case for mysql and another for postgresql.